### PR TITLE
doc: API stability policy

### PR DESCRIPTION
**Pull request by jasnell**
re: https://github.com/jasnell/dev-policy/issues/37

Expanded language around the API stability policy

### CODE REVIEW

**Comment by arb** [comment]
While it isn't what I _want_ the policy to be, it is much clearer than it was which is good.

File: README.md [hunk 1]
```diff
@@ -288,11 +288,27 @@ Any API *addition* to either the Node.js Core Library API or Application Binary
...
+### Implicit vs. Explicit API Stability
 
-APIs and default behaviors in the Node.js Core Library, Application Binary Interface, Dependencies and binary abstractions must not change within LTS Releases unless the change is required to address a critical security update.  
+The Node.js Core Library API, Application Binary Interface and Binary Abstraction layer each expose methods, events, behaviors and default values that together comprise the implicit Node.js "Public API". This "Public API" is used  by application and module developers and changes will have direct impact on the Community.
+
+Methods exposed by the API can have either strongly or weakly types input parameters and return values. Methods may also throw any number of exceptions or trigger certain sequences of events or side effects. When such types, exceptions, events or side-effects are *documented*, they become part of the Node.js *Explicit API*.
```
Comment (Fishrock123) [inline README.md]: >  weakly types input parameters

typed

File: README.md [hunk 2]
```diff
@@ -288,11 +288,27 @@ Any API *addition* to either the Node.js Core Library API or Application Binary
...
+
+Likewise, when the ordering of events triggered by Node.js, the default values of optional input parameters and constants, or the type and structure of return values and event payloads are *documented*, they become part of the *Explicit API*.
+
+*Undocumented* types, return values, exceptions, events and side effects are considered to be part of the *Implicit API*.
+
+The ecosystem of application and module developers must be ale to rely on the stability of both *Implicit* and *Explicit* APIs as much as possible. Changes to these APIs that are made prior to their inclusion within a Release are generally straightforward and can be made with appropriate review. Changes after the APIs have been included within a Release, however, require Collaborators to be much more careful.
```
Comment (Fishrock123) [inline README.md]: > developers must be ale to rely

able

File: README.md [hunk 3]
```diff
@@ -288,11 +288,27 @@ Any API *addition* to either the Node.js Core Library API or Application Binary
...
+
+Backwards incompatible changes to *Explicit* APIs that have previously been included in a Release *must* follow a clear and predictable deprecation process in which the existing documented API: (a) is clearly marked for deprecation and (b) can change only after the next *semver-major* version increase.
+
+Backwards incompatible changes to *Implicit* APIs that have been included in a Release *should* be handled the same as *Explicit* API changes. However, exception to this rule can be allowed if: (a) the proposed change can be reliably demonstrated to have minimal impact *in practice*, (b) the proposed change is intended to reconcile *implemented* behavior with *documented* behavior (i.e. the documentation says one thing but the code does another), or (c) it is clear that the Implicit API being modified had originally been intended for internal use as part of the underlying implementation as opposed to being targeted at developers.
```
Comment (Fishrock123) [inline README.md]: > (b) the proposed change is intended to reconcile _implemented_ behavior with _documented_ behavior (i.e. the documentation says one thing but the code does another)

Not always. See `Util.isObject` discussion in io.js - the change is too large to make to fix the function, so instead, the docs were fixed.

---

Also, there may need to be another case here: when something never actually worked to begin with (and wasn't documented).

File: README.md [hunk 4]
```diff
@@ -288,11 +288,27 @@ Any API *addition* to either the Node.js Core Library API or Application Binary
...
+
+That said, changes to any of the APIs should not be made lightly as there is no reliable means of determining in advance the impact any change may have on existing applications and modules. Collaborators must use their best judgement to determine whether any given change is "technically backwards incompatible but in practice should not be", then approach landing the change accordingly.
```
Comment (Fishrock123) [inline README.md:309]: > as there is no reliable means of determining in advance the impact any change may have on existing applications and modules.

Yet. We should aim to do some sort of ecosystem tests in the future.

**Comment by jasnell** [comment]
@arb nothing is set in stone. how would you change it?

File: README.md (hunk 4 above)
Comment (Fishrock123) [inline README.md:309]: Speaking of which, rust has recently started doing this; maybe we can take a cue from there? https://gist.github.com/brson/a30a77836fbec057cbee

See also: http://blog.rust-lang.org/2015/04/03/Rust-1.0-beta.html

Comment (jasnell) [inline README.md:309]: +1.. absolutely agreed. I've actually got an outstanding todo on this for an npm module test suite we've developed. Hoping to get back to that very soon.

**Comment by jasnell** [comment]
@Fishrock123 better after the new commit?

**Comment by arb** [comment]
I agree with @cjihrig - if you are using a public API method in an undocumented or unexpected way, that's on you when it breaks. There shouldn't be a major bump when you add validation or change the function so it no longer works the undocumented way. There is a lot of code that lingers around because people started using something wrong and now it's in there for good... 

Seems like that is not the prevailing opinion though.

**Comment by Fishrock123** [comment]
Yeah that's better.

**Comment by jasnell** [comment]
@arb... I tend to agree with both points of view which is why this attempts to add some wiggle room for exceptional cases.

**Comment by mikeal** [comment]
> if you are using a public API method in an undocumented or unexpected way, that's on you when it breaks

It's easy to say things like this but the reality is that if enough people are using/depending on an API, documented or not, we can't break it. If a change breaks a significant portion of the ecosystem it won't make it in, there's no moral or technical high ground we can find that justifies breaking the ecosystem. This is why @chrisdickinson has been trying to find ways of creating truly "internal only" modules so that we don't continue to put ourselves in a position where we have to support what was meant to be a private internal API.

**Comment by Fishrock123** [comment]
(Speaking of which, io.js now has support for internal modules.)

**Comment by jasnell** [comment]
Merging this change. Keep in mind that this can be iterated on further.

File: README.md [hunk 5]
```diff
@@ -288,11 +288,27 @@ Any API *addition* to either the Node.js Core Library API or Application Binary
...
+
+Backwards incompatible changes to *Explicit* APIs that have previously been included in a Release *must* follow a clear and predictable deprecation process in which the existing documented API: (a) is clearly marked for deprecation and (b) can change only after the next *semver-major* version increase.
+
+Backwards incompatible changes to *Implicit* APIs that have been included in a Release *should* be handled the same as *Explicit* API changes. However, exception to this rule can be allowed if: (a) the proposed change can be reliably demonstrated to have minimal impact *in practice*, (b) the proposed change is intended to reconcile *implemented* behavior with *documented* behavior (i.e. the documentation says one thing but the code does another and either the code or documentation need to be updated), (c) it is clear that the Implicit API being modified had originally been intended for internal use as part of the underlying implementation as opposed to being targeted at developers, or (d) it can be demonstrated that the Implicit API never worked correctly to begin with.
```
Comment (chrisdickinson) [inline README.md:307]: Nits, belatedly:

> (b) the proposed change is intended to reconcile _implemented_ behavior with _documented_ behavior (i.e. the documentation says one thing but the code does another and either the code or documentation need to be updated)

Documentation is a derivation of code, not the other way around. Changing code to match docs – especially if the change causes breakage – should follow the same process as explicit API changes. IOW: Changing docs to match behavior stands to break nobodies code, changing code to match docs _can_ break code – which should be avoided without a deprecation cycle.

> (c) it is clear that the Implicit API being modified had originally been intended for internal use as part of the underlying implementation as opposed to being targeted at developers

I'm thinking here of things like `_readableState`, or `_events` within EventEmitter – those things, while private implementation details, are so widely used throughout the ecosystem that any _change_ to those should probably be preceded by a lengthy deprecation cycle (or polyfilling those features back in.)

**Comment by chrisdickinson** [comment]
LGTM, with a few nits. 

@arb: the end result is the same – if you're using implicit APIs, you may eventually have to change your code. Core has to move more slowly in order to get to that point, though, because _so_ much (otherwise well-intentioned) code relies on those implicit APIs, and those dependencies can be incredibly distant from any given user's control. Deprecation cycles give the community the chance to identify problem modules in their dep trees and put pressure on those modules to update, or to create an alternative to that module.

**Pull request merged**